langroid/agent/chat_agent.py

```diff
@@ -581,6 +644,33 @@ def enable_message(
             self.system_tool_format_instructions = self.tool_format_rules()
         self.system_tool_instructions = self.tool_instructions()
 
+    def __getitem__(self, output_type: type) -> Self:
+        """
+        Returns a (shallow) copy of `self` with a forced output type.
+        As native function calls may be generated even with a forced
+        output format, and these may not necessarily belong to the
+        output type, we additionally disable the LLM's native function
+        calling mechanism.
+        """
+        clone = copy.copy(self)
+        if not any(issubclass(output_type, t) for t in [ToolMessage, BaseModel]):
+
+            class OutputFormat(BaseModel):
+                value: output_type  # type: ignore
+
+            output_type = OutputFormat
+
+        clone.output_format = output_type
+        if issubclass(output_type, ToolMessage):
+            clone.enabled_requests_for_inference = {
+                output_type.default_value("request")
+            }
+        if self.config.use_functions_api:
```

Comment:
If an agent with a strict output format should be able to produce a function call, `output_format` must be a `ToolMessage` and `output_format` should be the only usable tool. Using the native OpenAI tool calling capabilities doesn't ensure this, but we could choose to permit both structured outputs and arbitrary tool calls with a flag.

Reply:
maybe mark this as a TODO in the code to visit later.